lib/galaxy/files/uris.py

```diff
@@ -0,0 +1,168 @@
+
+from galaxy.exceptions import (
+    AdminRequiredException,
+    ConfigDoesNotAllowException,
+)
+from galaxy.util import (
+    DEFAULT_SOCKET_TIMEOUT,
+    get_charset_from_http_headers,
+    stream_to_open_named_file,
+    unicodify,
+)
+
+if TYPE_CHECKING:
+    from galaxy.files import ConfiguredFileSources
+
+
+log = logging.getLogger(__name__)
+
+
+def stream_url_to_str(
+    path: str, file_sources: Optional["ConfiguredFileSources"] = None, prefix: str = "gx_file_stream"
+) -> str:
+    tmp_file = stream_url_to_file(path, file_sources=file_sources, prefix=prefix)
+    try:
+        with open(tmp_file, "r") as f:
+            return f.read()
+    finally:
+        os.remove(tmp_file)
+
+
+def stream_url_to_file(
+    path: str, file_sources: Optional["ConfiguredFileSources"] = None, prefix: str = "gx_file_stream"
+) -> str:
+    temp_name: str
+    if file_sources and file_sources.looks_like_uri(path):
+        file_source_path = file_sources.get_file_source_path(path)
+        with tempfile.NamedTemporaryFile(prefix=prefix, delete=False) as temp:
+            temp_name = temp.name
+        file_source_path.file_source.realize_to(file_source_path.path, temp_name)
+    elif path.startswith("base64://"):
+        with tempfile.NamedTemporaryFile(prefix=prefix, delete=False) as temp:
+            temp_name = temp.name
+            temp.write(base64.b64decode(path[len("base64://") :]))
+            temp.flush()
+    else:
+        page = urllib.request.urlopen(path, timeout=DEFAULT_SOCKET_TIMEOUT)  # page will be .close()ed in stream_to_file
+        temp_name = stream_to_file(page, prefix=prefix, source_encoding=get_charset_from_http_headers(page.headers))
+    return temp_name
+
+
+def stream_to_file(stream, suffix="", prefix="", dir=None, text=False, **kwd):
+    """Writes a stream to a temporary file, returns the temporary file's name"""
+    fd, temp_name = tempfile.mkstemp(suffix=suffix, prefix=prefix, dir=dir, text=text)
+    return stream_to_open_named_file(stream, fd, temp_name, **kwd)
+
+
+IpAddressT = Union[ipaddress.IPv4Address, ipaddress.IPv6Address]
+IpNetwrokT = Union[ipaddress.IPv4Network, ipaddress.IPv6Network]
+IpAllowedListEntryT = Union[IpAddressT, IpNetwrokT]
+
+
+def validate_uri_access(uri: str, is_admin: bool, ip_allowlist: List[IpAllowedListEntryT]) -> None:
+    """Perform uniform checks on supplied URIs.
+
+    - Prevent access to local IPs not found in ip_allowlist.
+    - Don't allow non-admins to access file:// URIs.
+    """
+    validate_non_local(uri, ip_allowlist)
+    if not is_admin and uri.lstrip().startswith("file://"):
+        raise AdminRequiredException()
+
+
+def validate_non_local(uri: str, ip_allowlist: List[IpAllowedListEntryT]) -> str:
+    # If it doesn't look like a URL, ignore it.
+    if not (uri.lstrip().startswith("http://") or uri.lstrip().startswith("https://")):
+        return uri
+
+    # Strip leading whitespace before passing url to urlparse()
+    url = uri.lstrip()
```

Comment:
I would move this line to the beginning to simplify the condition a bit, not a big deal though since it has been like that for a long time

Reply:
I thought the same thing but then realized it would change the behavior slightly since we return the original URI without it being stripped if it isn't a HTTP URI. I did catch on this though too...

Reply:
Ahh I see! :+1: